### ISSUE TITLE
Enable using property mapping methods that contain @MappingTarget parameters

### ISSUE THREAD

**Issue by agudian** [mapstruct/mapstruct#160]
Mapping methods that contain an ```@MappingTarget``` parameter should be usable as property mapping methods.

The generated code should cover the following cases:
* The ```@MappingTarget``` method could return void or the target type. In case the target type is returned, we should call the setter with that value.
* The target property may be null. I guess we should hand in a new instance of the target type (call constructor or a factory method).

Any thoughts?

**Comment by sjaakd** [comment]
@agudian do you see time to work on this? Shall I give it a try?

**Comment by agudian** [comment]
How could I stop you? You're on fire lately! :smile: so don't hold back!

**Comment by sjaakd** [comment]
Thanks.. Doing some early-spring-cleaning :smile:

**Comment by sjaakd** [comment]
Ok, lets start out with an example.

```java
public class Game {

    private Player player;

    public Player getPlayer() {
        return player;
    }

    public void setPlayer(Player player) {
        this.player = player;
    }
}

public class Player {

    private String name;

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    } 
}

public class TargetGame {

    private TargetPlayer player;

    public TargetPlayer getPlayer() {
        return player;
    }

    public void setPlayer(TargetPlayer player) {
        this.player = player;
    }

}

public class TargetPlayer {

    private String name;

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }
}

@Mapper
public interface ExampleMapper {

    void mapGame( Game game, @MappingTarget TargetGame targetGame );

    void mapPlayer( Player player, @MappingTarget TargetPlayer targetPlayer );

    TargetPlayer mapPlayer( Player player );
}
```

So the result (I think) should be:

```java
public class ExampleMapperImpl1 implements ExampleMapper {

    @Override
    public void  mapGame(Game game, TargetGame targetGame)  {
        if ( game == null ) {
            return;
        }

        mapPlayer( game.getPlayer(), targetGame.getPlayer() );
    }

    @Override
    public void mapPlayer(Player player, TargetPlayer targetPlayer)  {
        if ( player == null ) {
            return;
        }

        targetPlayer.setName( player.getName() );
    }

    @Override
    public TargetPlayer mapPlayer(Player player)  {
        if ( player == null ) {
            return null;
        }

        TargetPlayer targetPlayer = new TargetPlayer();

        targetPlayer.setName( player.getName() );

        return targetPlayer;
    }
}
```

So, we should replace the property setter by: `mapPlayer( game.getPlayer(), targetGame.getPlayer() );`. I think we should do this regardless on whether `mapPlayer` returns the target or void to keep it simple.

I think that from a `@MappingTarget` annotated method, for a property mapping we should prefer a `@MappingTarget` annotated method when present. Note there could be multiple of them (switching the target with the source, with or without a non-void return. I think this should be covered in the filtering mechanism (as extra filter), so I'll wait on  #385 / #391.  :smile: 

@agudian , @gunnarmorling WDYT?

**Comment by agudian** [comment]
> I think that from a @MappingTarget annotated method, for a property mapping we should prefer a @MappingTarget annotated method when present.

Exactly, and the same needs to go for the opposite case, when you have a non-mapping-target method to generate and both variants are presesent for a property mapping method. (In my projects, we have mapping methods in both flavours for all the dto-to-entity constellations)

I have one more thing to add to your example, and that is the handling of a non-null-source property and a null target property:
```java
    @Override
    public void  mapGame(Game game, TargetGame targetGame)  {
        if ( game == null ) {
            return;
        }

        if ( targetGame.getPlayer() == null ) { // we could even leave that stuff out if getPlayer() is annotated with an annotation that has the simple (!) name @NonNull or @Nonnull or @NotNull
            targetGame.setPlayer( new Player() ); // or factory method, but you know where I'm headed...
        }
        mapPlayer( game.getPlayer(), targetGame.getPlayer() );
    }
```

Because that stuff can happen in reality and it would only then be possible to use a mapping-target method as property mapping method in a non-mapping-target method (in case there is no alternative non-mapping-target property mapping method defined that would otherwise have been preferred)

**Comment by sjaakd** [comment]
> Because that stuff can happen in reality and it would only then be possible to use a mapping-target method as property mapping method in a non-mapping-target method (in case there is no alternative non-mapping-target property mapping method defined that would otherwise have been preferred)

IIUC, an 'update' method must behave as a 'create' method if it is called by a 'create' method? I'm not sure on this one. Would't it be better that update methods only call other update methods and create methods would only call other create methods? So, keep these structures totally separated?

In addition there's the question what to do when you still have a defined source object, but no defined target object. I would say that we (initially) stop as well. In a later phase we might want to add a strategy thing there. Something like an (name needs to be improved) `MappingTargetStrategy: CreateTargets, Stop.`

So, `Stop` would look like this:
```java
   @Override
    public void  mapGame(Game game, TargetGame targetGame)  {
        if ( game == null ) {
            return;
        }

        if ( targetGame.getPlayer() != null ) {
           mapPlayer( game.getPlayer(), targetGame.getPlayer() );
        }
    }
```

And `CreateTargets` would look exactly like your code:
```java
    @Override
    public void  mapGame(Game game, TargetGame targetGame)  {
        if ( game == null ) {
            return;
        }

        if ( targetGame.getPlayer() == null ) {
            targetGame.setPlayer( new Player() ); // or factory method, but you know where I'm headed...
        }
        mapPlayer( game.getPlayer(), targetGame.getPlayer() );
    }
```

BTW: I'm not sure that we should already incorporate `NotNull` behavior, although, of course, we need to take simple types into consideration.

**Comment by agudian** [comment]
> BTW: I'm not sure that we should already incorporate NotNull behavior, although, of course, we need to take simple types into consideration.

Of course not, that's just and idea that hit me while typing it, and as it would affect other spots as well, we should think about that in a different issue.

**Comment by sjaakd** [comment]
Done. first step merged.
